### ISSUE TITLE
Idle Quod Libet consumes 2…3% of CPU.

### ISSUE THREAD

**Issue by van-de-bugger** [quodlibet/quodlibet#4620]
Steps to reproduce
------------------
1. Start Quod Libet.
2. Wait until Quod Libet finishes scanning the music library.
3. Make sure Quod Libet is idle: music is **not** playing.
4. Start a system monitor (Gnome System Monitor, `top`, whatever else), look at the CPU load.

Expected Output
---------------
`quodlibet` process is in Sleeping state with no CPU consumption.

Actual Output
-------------
`quodlibet` process is in Running state consuming 2…3% of CPU (as reported by Gnome System Monitor).

```
$ top -b -n 1 -p `pgrep quodlibet`

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
2340167 vdb       20   0 1996300 921188  59764 S  13.3   2.8  22:57.35 quodlibet
```

**Note:** Gnome System Monitor reports 100% load if *all CPU cores* are fully loaded. `top` program reports CPU load *per core*, it reports 100% load if *one CPU core* is fully loaded. On my quad core system, `top` reports that Quod Libet consumes 10…13% *of a core*.

Test System
-----------

Quod Libet: v4.6.0
OS: Fedora 38
Desktop: Gnome 44.10 + Xorg

Additional Information
----------------------

It seems the initial (aka main) thread of Quod Libet consumes CPU:
```
$ top -b -n 1 -H -p `pgrep quodlibet`

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
2340167 vdb       20   0 1996300 921188  59764 S  13.3   2.8  23:50.08 quodlibet
2340169 vdb       20   0 1996300 921188  59764 S   0.0   2.8   0:00.00 pool-spawner
2340170 vdb       20   0 1996300 921188  59764 S   0.0   2.8   0:00.00 gmain
2340171 vdb       20   0 1996300 921188  59764 S   0.0   2.8   0:00.03 gdbus
2340331 vdb       20   0 1996300 921188  59764 S   0.0   2.8   0:00.04 flacparse1:sink
2340338 vdb       20   0 1996300 921188  59764 S   0.0   2.8   0:00.02 aqueue:src
2340339 vdb       20   0 1996300 921188  59764 S   0.0   2.8   0:00.01 typefind:sink
2350860 vdb       20   0 1996300 921188  59764 S   0.0   2.8   0:00.25 threaded-ml
```

The problem may be not very serious for a desktop system running on AC power, but it is a noticeable load on a battery-powered laptop.

**Comment by declension** [comment]
Thanks for the detailed report. What plugins have you got enabled?

**Comment by van-de-bugger** [comment]
Enabled plugins:

*   Advanced Preferences
*   Application Information
*   Change Language
*   Gnome Search Provider
*   MPRIS D-Bus Support
*   Random Album Playback
*   Song Nitifications
*   Tray Icon
*   View Lyrics

Not sure I really need all of them. I will try to disable plugings one by one and see if it affects CPU load.

**Comment by declension** [comment]
Thanks. To gauge the problem a bit, could you share roughly how many songs and directories you have being watched (and on what kind of disk / network)?

**Comment by van-de-bugger** [comment]
~ 37000 files in ~ 6700 directories, ~ 7.9 TB in Ext4 filesystem, located on a hard drive (7200 RPM, Toshiba MG06ACA) and shared via NFS. The server is AMD E1-6010 @ 1.35 GHz with 8 GiB RAM. The network is gigabit ethernet. The client is Intel Core i5-2400 @ 3.1GHz with 32 GiB RAM. Typically, the client is the only system on the network (not counting the server, of course).

**Comment by declension** [comment]
Ahhh, it's _always_ NFS (or similar) when we get these types of issues :sweat_smile: 

I think watching network files in Linux is inherently problematic, but I have no proof exactly. Also, it should at least warn if this might be the case I guess. Happy for ideas here


Either way, updating ticket title now we have more specific reproduction criteria

**Comment by van-de-bugger** [comment]
Oops, my mistake: wrong total size of the library (in bytes), it is actually 10 times less. I edited the message, now it shows the correct size.

I'll also try to move the library to a local drive to see the difference.

**Comment by declension** [comment]
> Err... I have multiple computers (desktops + media center), I want the library to be available on each system, so putting it in a shared location is an obvious and typical scenario, right? Would you recommend another approach or another sharing method other than NFS? Is Samba better for this purpose?

Yes agree this is the best way of doing it, just have noticed it also correlates strongly with performance issues. Samba probably similar, but would be good to have actual data on this! 

> I'll also try to move the library to a local drive to see the difference in CPU load.

Yes, also a very useful data point (though plenty of anecdata in this setup already I guess)

**Comment by van-de-bugger** [comment]
I copied about half (I just do not have enough space locally to copy entire library) of my music library to a local drive. With the local library, Quod Libet spends 0.55…1.00% of CPU, and CPU load *does not depend* on  "Watch for directories for changes" setting. 

However, I would like to note:
1. QuodLibet is Running and consumes CPU cycles. 
2. I don't think NFS is the cause of high CPU usage.

I did an experiment: I wrote a sample test program in C++ that uses the inotify interface to monitor files and directories.

Few notes on inotify:
* inotify is not very convenient to monitor a directory tree: you have to monitor *every directory* of the tree.
* I had some issues with renaming: I get MOVED_FROM notifications, but never see MOVED_TO notifications.
* In other aspects inotify works quite well, *even if the files on an NFS share*.

My test program takes ~ 40 seconds to initialize (add watches for all the ~ 6700 directories on NFS share), but then it is in Sleep mode (and do not consume CPU cycles at all) and wakes up only briefly to print a message about observed activity. Of course, the number of wakeups depends on the subscriptions the test program made: if the test program subscribes to all events, it is notified if someone reads the watched directories and files. But if the test program subscribes to `IN_ONLYDIR | IN_CREATE | IN_CLOSE_WRITE | IN_DELETE | IN_DELETE_SELF | IN_MOVED_FROM | IN_MOVED_TO`, it sleeps until someone actually changes the library.

Thus, "Watching network libraries consumes 2%…3% of CPU" is true, but I guess the problem is not in the network library. How QoudLibet watches the library for changes?
